File: backend/src/application/commands/upsert_library_item.py
from __future__ import annotations

from dataclasses import dataclass

from domain.models import LibraryItem
from domain.repositories import LibraryRepository


@dataclass(frozen=True)
class UpsertLibraryItemCommand:
    item: LibraryItem


class UpsertLibraryItemHandler:
    def __init__(self, repository: LibraryRepository) -> None:
        self._repository = repository

    def handle(self, command: UpsertLibraryItemCommand) -> LibraryItem:
        return self._repository.upsert(command.item)

File: frontend/src/application/commands/UpsertLibraryItem.ts
import { MangaSeries } from '@domain/entities/MangaSeries';
import { LibraryRepository } from '@domain/repositories/LibraryRepository';

export interface UpsertLibraryItemCommand {
  item: MangaSeries;
}

export class UpsertLibraryItem {
  constructor(private repository: LibraryRepository) {}

  handle(command: UpsertLibraryItemCommand): Promise<MangaSeries> {
    return this.repository.upsert(command.item);
  }
}
